src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ComponentWrapper.java

```diff
@@ -0,0 +1,150 @@
+	 * Ensures up-to-date component data retrieval and modification on the {@link ItemStack} of the {@link ItemSource} .
+	 * @param itemSource The {@link ItemSource} representing the original source of the {@link ItemStack}.
+	 */
+	public ComponentWrapper(ItemSource itemSource) {
+		this.itemSource = itemSource;
+		this.itemStack = itemSource.getItemStack();
+		this.component = this.getComponent(itemStack.getItemMeta());
+	}
+
+	/**
+	 * Constructs a {@link ComponentWrapper} that only references to a component.
+	 */
+	public ComponentWrapper(T component) {
+		this.component = component;
+		this.itemSource = null;
+		this.itemStack = null;
+	}
+
+	/**
+	 * Returns the current component.
+	 * If this {@link ComponentWrapper} was constructed with an {@link ItemSource}, the component is retrieved from
+	 * the stored item. Otherwise, the stored {@link #component}.
+	 */
+	public T getComponent() {
+		if (itemSource != null) {
+			assert itemStack != null;
+			return this.getComponent(itemStack.getItemMeta());
+		}
+		return component;
+	}
+
+	/**
+	 * Returns the {@link ItemStack} associated with this {@link ComponentWrapper}, if available.
+	 */
+	public @Nullable ItemStack getItemStack() {
+		return itemStack;
+	}
+
+	/**
+	 * Returns the {@link ItemSource} the {@link ItemStack} is sourced from.
+	 */
+	public @Nullable ItemSource getItemSource() {
+		return itemSource;
+	}
+
+	/**
+	 * Returns the {@link Converter} used to extract the component from the {@link ItemMeta}.
+	 */
+	protected abstract T getComponent(ItemMeta itemMeta);
+
+	/**
+	 * Returns the {@link BiConsumer} that updates the component on the {@link ItemMeta}.
+	 */
+	protected abstract void setComponent(ItemMeta itemMeta, T component);
+
+	/**
+	 * Apply the current {@link #component} to the {@link #itemSource}.
+	 */
+	public void applyComponent() {
+		applyComponent(getComponent());
+	}
+
+	/**
+	 * Apply a new {@code component} or {@link #component} to the {@link #itemSource}.
+	 */
+	public void applyComponent(@NotNull T component) {
+		if (itemSource == null)
+			return;
+		assert itemStack != null;
+		ItemMeta itemMeta = itemStack.getItemMeta();
+		setComponent(itemMeta, component);
+		itemSource.setItemMeta(itemMeta);
+	}
+
+	/**
+	 * Edit {@link #component} via {@link Consumer}.
+	 * @param consumer The {@link Consumer} to edit the component.
+	 */
+	public void editComponent(Consumer<T> consumer) {
+		T component = getComponent();
+		consumer.accept(component);
+		applyComponent(component);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (!(obj instanceof ComponentWrapper<?> other))
+			return false;
+		boolean relation = true;
+		if (this.itemStack != null && other.itemStack != null)
+			relation = this.itemStack.equals(other.itemStack);
+		relation &= this.getComponent().equals(other.getComponent());
+		return relation;
+	}
+
+	/**
+	 * Get a clone of this {@link ComponentWrapper}.
+	 */
+	public abstract ComponentWrapper<T> clone();
```

Comment:
If we're calling this clone, it might be good to ensure the conventions are followed: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Object.html#clone()

Additionally, consider implementing Cloneable